### PR TITLE
Fix compilation issues in validation filter

### PR DESCRIPTION
**Pull request by hasuniea**
### Purpose
Fix compilation issues in validation filter.

### Issues
https://github.com/wso2/product-microgateway/issues/717

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 1]
```diff
@@ -1,287 +1,286 @@
...
-//                 }
-//             }
-//         }
-//         else {}
-
-//     }
-//     return true;
-// }
+        //payload can be of type json or error
+        if(payload is json) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]: Negate the if block (Return if payload is not a json and continue)

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 2]
```diff
@@ -1,287 +1,286 @@
...
+            //do the validation if only there is a payload and a model available. prevent validating error
+            //responses sent from the filterRequest if the request is invalid.
+            if (model != null && payload != null && payload.fault == null && model is json && models is json) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]: Same as above, negate the if block

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 3]
```diff
@@ -1,287 +1,286 @@
...
+                    int errorCode = <int>context.attributes[ERROR_CODE];
+                    //changing the response
+                    response.statusCode = statusCode;
+                    response.setContentType(APPLICATION_JSON);
+                    //creating a new payload which is having the error message
+                    json newPayload = { fault: {
+                        code: errorCode,
+                        message: errorMsg,
+                        description: errorDescription
+                    } };
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]: How about,                   
```ballerina
json newPayload = {
    fault: {
        code: errorCode,
        message: errorMsg,
        description: errorDescription
    }
};
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 4]
```diff
@@ -1,287 +1,286 @@
...
+                                  modelName = replaceModelPrefix(modelReference);
+                                  if (models.modelName != null) {
+                                      model = models.modelName;
+                                  }
+                               } else {
+                                         //getting inline model defined under the schema
+                                           model = schema;
+                               }
+                           }
+                             if (schema.TYPE != null) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]: If block is not formatted properly

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 5]
```diff
@@ -1,287 +1,286 @@
...
+                        if (swagger.PATHS.i.requestMethod.RESPONSES.responseStatusCode.SCHEMA.TYPE != null) {
+                            isType = true;
+                            pathType = <@untainted>  swagger.PATHS.i.requestMethod.RESPONSES.responseStatusCode.
+                            SCHEMA.TYPE.toString();
+                        }
+               } else {
+                   if (swagger.PATHS.i.requestMethod.RESPONSES.responseStatusCode.CONTENT.APPLICATION_JSON.
+                       SCHEMA != null) {
+                       json|error schema = swagger.PATHS.i.requestMethod.RESPONSES.responseStatusCode.CONTENT.APPLICATION_JSON.SCHEMA;
+                        if (schema is json) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]: We can negate the if block and continue safely to simplify the logic

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 6]
```diff
@@ -1,287 +1,286 @@
...
+                                modelName = replaceModelPrefix(modelReference);
+                                if (models.modelName != null) {
+                                    model = models.modelName;
+                                }
+                         } else {
+                                //getting inline model defined under the schema
+                                model = swagger.PATHS.i.requestMethod.RESPONSES.responseStatusCode.SCHEMA;
+                         }
+                   }
+                        if (swagger.PATHS.i.requestMethod.RESPONSES.responseStatusCode.SCHEMA.TYPE != null) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]: If block not formatted properly

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 7]
```diff
@@ -1,287 +1,286 @@
...
+            models = swagger.components.schemas;
+        } else if (swagger.definitions != null) {//getting schemas from a swagger 2.0 version file
+            models = swagger.definitions;
+        }
+        map<json> pathMap = <map<json>>pathKeys;
+        foreach var i in pathMap {
+           if (requestPath == i && models is json) {
+               string modelReference;
+               if (swagger.PATHS.i.requestMethod.RESPONSES.responseStatusCode.SCHEMA != null) {
+                   if (swagger.PATHS.i.requestMethod.RESPONSES.responseStatusCode.SCHEMA.ITEMS != null) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]: Please reformat below if-else blocks. All below blocks are formatted with more than 4 spaces

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 8]
```diff
@@ -1,287 +1,286 @@
...
+        if (swagger.components.schemas != null) {//getting the schemas from a swagger 3.0 version file
+            models = swagger.components.schemas;
+        } else if (swagger.definitions != null) {//getting schemas from a swagger 2.0 version file
+            models = swagger.definitions;
+        }
+        map<json> pathMap = <map<json>>pathKeys;
+        foreach var i in pathMap {
+           if (requestPath == i && models is json) {
+               string modelReference;
+               if (swagger.PATHS.i.requestMethod.RESPONSES.responseStatusCode.SCHEMA != null) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]: Can negate the if block and continue

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 9]
```diff
@@ -1,287 +1,286 @@
...
+        string modelName = "";
+        string responseStatusCode = response.statusCode.toString();
+        if (swagger.components.schemas != null) {//getting the schemas from a swagger 3.0 version file
+            models = swagger.components.schemas;
+        } else if (swagger.definitions != null) {//getting schemas from a swagger 2.0 version file
+            models = swagger.definitions;
+        }
+        map<json> pathMap = <map<json>>pathKeys;
+        foreach var i in pathMap {
+           if (requestPath == i && models is json) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]: Can negate the if block and continue

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 10]
```diff
@@ -1,287 +1,286 @@
...
-//                     sendErrorResponse(caller, request, filterContext);
-//                     return false;//avoid sending the invalid request to the backend by returning false.
-//                 }
-//             }
-//         } else {}
-//     }
-//     return true;
-// }
+        //payload can be of type json or error
+        if(payload is json && model is json && models is json) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]: Can simplify the nested if blocks by negating the condition

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 11]
```diff
@@ -1,287 +1,286 @@
...
+            if (models is json) {
+                foreach var i in pathMap {
+                    if (requestPath == i) {
+                        json|error parameters = swagger.PATHS.i.requestMethod.PARAMETERS;
+                        //go through each item in parameters array and find the schema property
+                        //go through each item in parameters array and find the schema property
+                        if (parameters != null) {
+                            map<json> para = <map<json>>parameters;
+                            //json[] para = <json[]>parameters;
+                                foreach var k in para {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]: Indentation issue

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 12]
```diff
@@ -1,287 +1,286 @@
...
+                            } else {
+                                //getting inline model
+                                model = k.SCHEMA;
+                            }
+                        } else {
+                            //getting inline model
+                            model = k.SCHEMA;
+                        }
+                    }
+                    } else {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]: Please correct the indentation in above block.

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 13]
```diff
@@ -1,287 +1,286 @@
...
+        isType = false;
+        APIConfiguration? apiConfig = apiConfigAnnotationMap[filterContext.getServiceName()];
+        json swagger = openAPIs[filterContext.getServiceName()];
+        printDebug(KEY_VALIDATION_FILTER, "The swagger content found in map : " + swagger.toString());
+        json|error model = {};
+        json|error models = {};
+        string modelName = "";
+        //getting all the keys defined under the paths in the swagger
+        pathKeys = swagger.PATHS;
+        //getting the methnted>  od of the request
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]:
```suggestion
        //getting the method of the request
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal [hunk 14]
```diff
@@ -1,287 +1,286 @@
...
-//                       //check whether there is a model available from the assigned model name
-//                        if (models[modelName] != null) {
-//                            model = models[modelName];
-//                        }
-//                    }
-//                 }
-//         }
+function doValidationFilterRequest(http:Caller caller, http:Request request, http:FilterContext filterContext, map<json> openAPIs)
+             returns boolean {
+    if (enableRequestValidation) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/validation_filter.bal]: How about returning if validation is not enabled?

**Comment by Rajith90** [comment]
@hasuniea , I will close this PR for now. Please re open and do the suggestions

**Pull request closed without merging**